vividus-plugin-mobitru/src/main/java/org/vividus/mobitru/client/MobitruFacadeImpl.java

```diff
@@ -54,10 +63,43 @@ public MobitruFacadeImpl(MobitruClient mobitruClient)
     @Override
     public String takeDevice(DesiredCapabilities desiredCapabilities) throws MobitruOperationException
     {
-        Waiter waiter = new DurationBasedWaiter(new WaitMode(waitForDeviceTimeout, RETRY_TIMES));
-        Device requestedDevice = new Device();
-        requestedDevice.setDesiredCapabilities(desiredCapabilities.asMap());
-        String capabilities = performMapperOperation(mapper -> mapper.writeValueAsString(requestedDevice));
+        Map<String, Object> capabilities = desiredCapabilities.asMap();
+        Device device;
+        Waiter deviceWaiter;
+        if (deviceSearchParameters.isEmpty() || Stream.of(UDID, "appium:udid", "deviceName", "appium:deviceName")
+                .anyMatch(capabilities::containsKey))
+        {
+            device = new Device();
+            device.setDesiredCapabilities(desiredCapabilities.asMap());
+            deviceWaiter = new DurationBasedWaiter(new WaitMode(waitForDeviceTimeout, RETRY_TIMES));
+        }
+        else
+        {
+            DeviceSearchParameters deviceSearchConfiguration = new DeviceSearchParameters(deviceSearchParameters,
+                    desiredCapabilities);
+            device = findDevices(deviceSearchConfiguration).get(0);
```

Comment:
if you are unable to get the first one and a few were found, does it make sense to peek next one and try retry acquiring?

Reply:
yes, it makes sense